
# TD1 - Knowledge Extraction from text
### *Nicolas AUDOUX - Thomas PAUL - Yana RAGOZINA*
### SI5 IA-ID




In this lab we will be focusing on performing a keyword/keyphrase extraction analysis (KPE) on a datasest of documents. Therefore, we will try to build and evaluate different document summaries  generated by different keyphrase extraction algorithms.

We will try to implement 3 different keyphrase extraction algorithms in order to analyse and compare their functionalities and performances.

We will use one of the datasets designed for automatic keyphrase extraction, *Inspec*, collecting 2000 abstract documents in English language in the domain of Computer Science.

As for the tested  KPE algorithms, we chose to analyse *PositionRank*, *SingleRank* and *TextRank*.

The performances will be evaluated with ROUGE framework (Recall-Oriented Understudy for Gisting Evaluation), designed to compare an automatically produced summary  by the algorithms against a reference.

### Imports

In [63]:
import pandas as pd
import pke
from rouge import Rouge
from os import listdir
from time import time
import spacy
nlp = spacy.load("en_core_web_sm")

In [64]:
# Defining constants
# pos and grammar for Position Rank
pos = {'NOUN', 'PROPN', 'ADJ'}
grammar = "NP: {<ADJ>*<NOUN|PROPN>+}"

# Extractors
position_rank_extractor = pke.unsupervised.PositionRank()
single_rank_extractor = pke.unsupervised.SingleRank()
text_rank_extractor = pke.unsupervised.TextRank()

rouge = Rouge()

In [65]:
def extract_keyphrases_score(extractor, doc, grammar=None, text_rank=None):
   # load the content of the document
   extractor.load_document(input=doc, language='en', normalization=None)

   # select the noun phrases up to 3 words as keyphrase candidates
   if grammar is not None:
      extractor.candidate_selection(grammar=grammar, maximum_word_number=3)
   else:
      extractor.candidate_selection()

   # weight the candidates using the sum of their word's scores that are
   # computed using random walk biaised with the position of the words
   # in the document. In the graph, nodes are words (nouns and
   # adjectives only) that are connected if they occur in a window of
   # 10 words.
   if text_rank:
      extractor.candidate_weighting(window=10, pos=pos, top_percent=0.33)
   else:
      extractor.candidate_weighting(window=10, pos=pos)

   # get the 10-highest scored candidates as keyphrases
   keyphrases = extractor.get_n_best(n=10)

   # compute rouge scores
   scores = rouge.get_scores(keyphrases[0][0], doc)

   return scores

In [66]:
def mean_all_scores(all_scores):
    mean_r_1, mean_p_1, mean_f_1 = 0, 0, 0
    mean_r_2, mean_p_2, mean_f_2 = 0, 0, 0
    mean_r_l, mean_p_l, mean_f_l = 0, 0, 0

    total_scores = len(all_scores)

    for scores in all_scores:
        mean_r_1 += scores['rouge-1']['r']
        mean_p_1 += scores['rouge-1']['p']
        mean_f_1 += scores['rouge-1']['f']

        mean_r_2 += scores['rouge-2']['r']
        mean_p_2 += scores['rouge-2']['p']
        mean_f_2 += scores['rouge-2']['f']

        mean_r_l += scores['rouge-l']['r']
        mean_p_l += scores['rouge-l']['p']
        mean_f_l += scores['rouge-l']['f']

    mean_r_1 /= total_scores
    mean_p_1 /= total_scores
    mean_f_1 /= total_scores

    mean_r_2 /= total_scores
    mean_p_2 /= total_scores
    mean_f_2 /= total_scores

    mean_r_l /= total_scores
    mean_p_l /= total_scores
    mean_f_l /= total_scores

    return mean_r_1, mean_p_1, mean_f_1, mean_r_2, mean_p_2, mean_f_2, mean_r_l, mean_p_l, mean_f_l

In [67]:
def get_scores(limitSize):
    all_scores_pr, all_scores_sr, all_scores_tr  = [], [], []
    duration_pr, duration_sr, duration_tr = 0, 0, 0

    dir = "Inspec/docsutf8/"
    directory = [dir+f for f in listdir(dir)][:limitSize]

    for i in directory:
        try:
            with open(i) as inspec_file:
                doc = inspec_file.read()
            print(f"Processing file {i}", end='\r')
        except:
            continue
        t1 = time()
        scores_pr = extract_keyphrases_score(position_rank_extractor, doc, grammar)
        t2=time()
        scores_sr = extract_keyphrases_score(single_rank_extractor, doc)
        t3=time()
        scores_tr = extract_keyphrases_score(text_rank_extractor, doc, text_rank=True)
        t4=time()

        duration_pr += t2-t1
        duration_sr += t3-t2
        duration_tr += t4-t3

        if scores_pr != 0:
            all_scores_pr.append(scores_pr[0])
        if scores_sr != 0:
            all_scores_sr.append(scores_sr[0])
        if scores_tr != 0:
            all_scores_tr.append(scores_tr[0])

    return all_scores_pr, all_scores_sr, all_scores_tr, duration_pr, duration_sr, duration_tr

In [68]:
def print_scores(scores,duration):
  mean_r_1, mean_p_1, mean_f_1, mean_r_2, mean_p_2, mean_f_2, mean_r_l, mean_p_l, mean_f_l = mean_all_scores(scores)
  print("Suration (in seconds) :",duration)
  print("Mean_r_1:", mean_r_1)
  print("Mean_p_1:", mean_p_1)
  print("Mean_f_1:", mean_f_1)

  print("Mean_r_2:", mean_r_2)
  print("Mean_p_2:", mean_p_2)
  print("Mean_f_2:", mean_f_2)

  print("Mean_r_l:", mean_r_l)
  print("Mean_p_l:", mean_p_l)
  print("Mean_f_l:", mean_f_l)

In [69]:
#Return the keyphrase of the document
def extract_keyphrases(extractor, doc, index, grammar=None, text_rank=None, pos=None):
    try:
        # load the content of the document
        extractor.load_document(input=doc, language='en', normalization=None)

        # select the noun phrases up to 3 words as keyphrase candidates
        if grammar is not None:
            extractor.candidate_selection(grammar=grammar, maximum_word_number=3)
        else:
            extractor.candidate_selection()

        if text_rank:
            extractor.candidate_weighting(window=10, pos=pos, top_percent=0.33)
        else:
            extractor.candidate_weighting(window=10, pos=pos)

        # get the 10-highest scored candidates as keyphrases
        keyphrases = extractor.get_n_best(n=10)

        if keyphrases is not None and keyphrases:
            res = f"{index} : {keyphrases[0][0]}"
        else:
            res = f"{index} : None"

    except Exception as e:
        # Handle specific exceptions if needed
        res = f"Error processing {index}: {str(e)}"

    # return the most predominant keyphrase
    return res

In [70]:
def get_keyphrases(limitSize):
    all_keyphrases_pr, all_keyphrases_sr, all_keyphrases_tr = [],[], []

    dir = "Inspec/docsutf8/"
    directory = [dir+f for f in listdir(dir)][:limitSize]

    for i in directory:
        try:
            with open(i) as inspec_file:
                doc = inspec_file.read()
            print(f"Processing file",i, end='\r')
        except:
            continue

        keyphrase_pr = extract_keyphrases(position_rank_extractor, doc, i, grammar)
        keyphrase_sr = extract_keyphrases(single_rank_extractor, doc,i)
        keyphrase_tr = extract_keyphrases(text_rank_extractor, doc,i,text_rank=True)
        
        all_keyphrases_pr.append(keyphrase_pr)
        all_keyphrases_sr.append(keyphrase_sr)
        all_keyphrases_tr.append(keyphrase_tr)

    return all_keyphrases_pr, all_keyphrases_sr, all_keyphrases_tr

In [ ]:
# extract keyphrases of the first 5 documents
all_keyphrases_pr, all_keyphrases_sr, all_keyphrases_tr = get_keyphrases(5)

In [76]:

print(all_keyphrases_pr)
print(all_keyphrases_sr)
print(all_keyphrases_tr)

['Inspec/docsutf8/100.txt : separate accounts', 'Inspec/docsutf8/1000.txt : universality', 'Inspec/docsutf8/1001.txt : conflict ibs', 'Inspec/docsutf8/1002.txt : selective representing-the idea', 'Inspec/docsutf8/1003.txt : lob']
['Inspec/docsutf8/100.txt : new entrants', 'Inspec/docsutf8/1000.txt : classical symbol systems', 'Inspec/docsutf8/1001.txt : content atomism', 'Inspec/docsutf8/1002.txt : selective representing-the idea', 'Inspec/docsutf8/1003.txt : human-like epistemic agents']
['Inspec/docsutf8/100.txt : new entrants', 'Inspec/docsutf8/1000.txt : universality', 'Inspec/docsutf8/1001.txt : content atomism', 'Inspec/docsutf8/1002.txt : realist conception', 'Inspec/docsutf8/1003.txt : epistemic authority']


When comparing the different keyphrases extracted by our algorithms, we notice that single rank and text rank are very similar (we have the same keyphrases for the first 5 documents). Position rank has different results for most of the texts (not the 1002). However, even if the keyphrases are different, there mostly part of the keys phrases contained in the folder keys. 
If we take the 100.txt doc, it is quite reasonable to pick *separate accounts* as a keyphrases since its is part of the title and the main subject of the doc. *new entrants* seems less coherent on this document. For the second document (1000.txt), we have a similar result: *Universality* is the main topic of the document whereas *classical symbol systems* is just an example. For those two example the result of the PositionRank seems to be the best.

In [ ]:
all_scores_pr, all_scores_sr, all_scores_tr, duration_pr, duration_sr, duration_tr = get_scores(100)

## 1. Position Rank

Position Rank extracts keyphrases by determining the importance of a word based on its position in the document.

It's an unsupervised algorithm that is decomposed like this :
1. Calculates the Term Frequency of a word(TF)
2. Adjusts the term frequency based on the length of the document (Document Length Normalization)
3. Assigns scores to words based on their positions within sentences. **Words in the beggining and end of sentences have higher scores.** (Sentence Position Score)
4. Combines the term frequency and sentence position scores to determine the overall importance of each word (Sentence Salience Score)
5. Extracts words that have the highest salience scores (Keyphrase Extraction)

In [73]:
print_scores(all_scores_pr,duration_pr)

Suration (in seconds) : 198.8953652381897
Mean_r_1: 0.02976155012249593
Mean_p_1: 0.8816666666666667
Mean_f_1: 0.057289591776589835
Mean_r_2: 0.011282417062540108
Mean_p_2: 0.73
Mean_f_2: 0.02213746087937282
Mean_r_l: 0.02976155012249593
Mean_p_l: 0.8816666666666667
Mean_f_l: 0.057289591776589835


## 2. SingleRank

SingleRank is a graph-based summarization algorithm. Therefore, it performs in the first place a graph representation of a given document where each sentence is represented as a node. Plus, this algorithm measures the degree of similarity between each pair of sentences of the document. Then, an importance degree (centrality) is assigned to each sentence, allowing a more precise analysis for building a summary. The further process of the summary generation can be seen as follows :

- Calculating the similarity between each pair of sentences in the document, often based on the content overlap
- Mapping the similar sentences as nodes to the graph where the measured similarities are used to build edges between the nodes.
- Measuring the degree of importance of each node (sentence) within the graph
- Ranking each sentence in descending order based on the previously calculated degree of importance (centrality)
- Generating the summary of the document based on the highest-ranking sentences to form the summary



In [74]:
print_scores(all_scores_sr,duration_sr)

Suration (in seconds) : 195.1198558807373
Mean_r_1: 0.03979474361445328
Mean_p_1: 0.9201666666666667
Mean_f_1: 0.07561336635076116
Mean_r_2: 0.018064838153112422
Mean_p_2: 0.8246666666666669
Mean_f_2: 0.03501314950770341
Mean_r_l: 0.03936969250975165
Mean_p_l: 0.9131666666666667
Mean_f_l: 0.07481193607998106


## 3. TextRank
TextRank is an algorithm that identifies keywords by assessing their significance within a connected graph. It functions by analyzing the relationships between words or phrases to determine their importance in the context of the overall text.
Here is he algorithm:
* Tokenization and part of speech tagging
* Reducing the number of words based on a syntactic filter (in our case we keep only noons propositions and adjectives)
* With all the remainig words are added to the graph and an edge is craeted for every words that co-occur in a window of N words (in our case, N=10)

At this point we have an undirected unweigth graph.

* Then a initial value of 1 is set for every vertice
* Finally a modify version of the PageRank algorithm is run to upgrade the vertice score.
The main idea behind this algorithm is to give more importance to a word which is linked by many others. Moreover a link to word which is linked by many other is more important than a link to word which is linked to only one word. This is the same algorithm used to rankes web pages. The only difference is that we also use a weight to each wich corresponds to the co-occurence score
* After that, we keep only a a third of our vertices which corresponds to the vertices which have the highest score.
* A post processing is done on the remainng vertices and if two words appears next to each other in the document a multi-word keyword is created.

In [75]:
print_scores(all_scores_tr,duration_tr)

Suration (in seconds) : 199.66671299934387
Mean_r_1: 0.03242190965438641
Mean_p_1: 0.9108333333333334
Mean_f_1: 0.062183449477588165
Mean_r_2: 0.013272283690595209
Mean_p_2: 0.79
Mean_f_2: 0.02591573531883978
Mean_r_l: 0.03226806350054026
Mean_p_l: 0.9075000000000002
Mean_f_l: 0.06188933183052934


## Conclusion

**Which algorithm got the best RED score?**

From the 3 keyphrase extraction algorithms (Position Rank, Single Rank and Text Rank), Single Rank has the best ROUGE score.



**How would you represent each document and its respective extracted key phrases in the form of a knowledge graph? What vocabulary would you use?**

Low-level : each document would have a graph with nodes representing the extracted keyphrases.

High-level : every documents would be represented as 1 node and would be linked by their predominant keyphrase extracted.

We can use the vocabulary of the extracted keyphrases of all the documents.


To represent this as a graph, we would have each document and each keyphrases as node. If a key phrase represents a document, we connect each other with a weighted link to measure the level of confidence. The vocabulary used can be all the extracted keyphrases. To have a more interesting graph, we could also compute the similarity between differents keyphrases to link them or link every keyphrases to a general topic.